### PR TITLE
dev/check: Add check for diff when running go mod tidy

### PR DESCRIPTION
**Pull request by indradhanush**
This commit adds an additional check for `go mod tidy`. If it produces
a diff in `go.mod` and / or `go.sum`, the pipeline should fail
instructing the user to commit the changes.

Inspired from #18973.

**Pull requests to fix go mod since this PR was created:**
- #19967
- #21447

We expect the build to fail when `go mod tidy` generates a diff: https://buildkite.com/sourcegraph/sourcegraph/builds/89560#af8318d8-4c2c-4d8c-9ead-3bbf64983bda/451-452

~Once #18973 is merged, I will rebase and a new build should then pass. :)~ Done.

### CODE REVIEW

**Comment by keegancsmith** [comment]
why do we want to enforce that go mod tidy has been run? Extra entries in our go.sum don't do anything bad, and there have been cases where go tools mutate go.mod/go.sum depending on which command is run. EG in the past you had to run `go test ./...` to ensure all go.mod/go.sum entries appear.

**Comment by indradhanush** [comment]
@keegancsmith I thought that maybe if we automate this as a check then `go.mod` and `go.sum` gets updated as and when new dependencies or existing ones are removed as part of a PR instead of someone in the team seeing linting errors in their local dev and then creating a PR out of their way as was the case in #18973. 

I wasn't aware of go tools mutating those files. If it appears that this additional check would increase complexity for dev work rather than reduce it, I'd be happy to close it. I will follow your lead on this. Let me know. :)

**Comment by keegancsmith** [comment]
> linting errors in their local dev and then creating a PR out of their way as was the case in #18973.

That sounds like a noisy lint check! 

> I wasn't aware of go tools mutating those files.

My knowledge may be very outdated from when go modules was beta. So hopefully they have improved this. Additionally we aren't doing as many fancy tricks as we used to with external tooling depending on our go.mod.

> If it appears that this additional check would increase complexity for dev work rather than reduce it, I'd be happy to close it.

I'm happy to try this out if you want to. We just need to be trigger happy to revert it if it leads to more frustration than the noisy linter :)

**Comment by indradhanush** [comment]
Yes, that makes sense. We can pilot it and would be happy to revert this if it makes our lives harder in the coming days. I noticed that #18973 was closed without being merged so I will have to dig into that issue first if I want to get this merged. :)

**Comment by indradhanush** [comment]
This fell through the cracks for me. I updated this with main. And now I don't see any new changes when running `go mod tidy` in this branch. :)

**Comment by indradhanush** [comment]
We've recently had this scenario where we had to create separate PRs just to update go modules. We could pilot this CI check and revert it if it ends up becoming too noisy and detrimental to developer productivity.

File: dev/check/go-mod-tidy.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+export PATH=$GOBIN:$PATH
+export GO111MODULE=on
+
+echo "--- go mod tidy"
+
+go mod tidy
+
+# Check if git diff contains go.mod and/or go.sum. In case the diff does not contain either of these files, `grep -c` will return a non-zero exit code leading to this script exiting as we `set -e` above. To avoid this, when `grep -c` returns no matches and fails we execute `true` to avoid a premature non-zero exit code. Detailed explanation is here:
+# https://unix.stackexchange.com/questions/330660/prevent-grep-from-exiting-in-case-of-nomatch
+readonly DIFF=$(git diff --name-only | grep -e 'go.mod' -e 'go.sum' -c || true)
```
Comment (rmmh) [inline dev/check/go-mod-tidy.sh:17]: maybe  `DIFF=$(git diff --name-only go.mod go.sum)` instead?

**Review by rmmh — APPROVED** [review]

File: dev/check/go-mod-tidy.sh (hunk 1 above)
Comment (efritz) [reply dev/check/go-mod-tidy.sh:17]:
```suggestion
readonly DIFF=$(git diff --name-only go.mod go.sum)
```

**Review by efritz — APPROVED** [review]

**Comment by indradhanush** [comment]
Link to failing build with uncommitted `go mod tidy` changes: https://buildkite.com/sourcegraph/sourcegraph/builds/98188#85c79fbf-8476-4f03-bfe1-359350a22902

And the latest passing build with the changes committed: https://buildkite.com/sourcegraph/sourcegraph/builds/98191#795e965c-1fff-4589-ac95-6a7cbdac6ac2

**Review by unknwon — APPROVED** [review]
Happy to experiment it and if it gets noisy we can just disable that.

**Pull request merged**